sys/net/network_layer/ng_ndp/ng_ndp.c

```diff
@@ -0,0 +1,950 @@
+#include "net/ng_netbase.h"
+#include "net/ng_netif/hdr.h"
+#include "net/ng_pktbuf.h"
+#include "random.h"
+#include "utlist.h"
+#include "thread.h"
+#include "vtimer.h"
+
+#include "net/ng_ndp.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static ng_pktqueue_node_t _pkt_nodes[NG_IPV6_NC_SIZE * 2];
+static ng_ipv6_nc_t *_last_router = NULL;   /* last router chosen as default
+                                             * router. Only used if reachability
+                                             * is suspect (i. e. incomplete or
+                                             * not at all) */
+
+/* random helper function */
+static inline uint32_t _rand(uint32_t min, uint32_t max)
+{
+    return (genrand_uint32() % (max - min)) + min;
+}
+
+static bool _handle_sl2a_opt(kernel_pid_t iface, ng_pktsnip_t *pkt,
+                             ng_ipv6_hdr_t *ipv6, uint8_t icmpv6_type,
+                             ng_ndp_opt_t *sl2a_opt);
+static bool _handle_tl2a_opt(kernel_pid_t iface, ng_pktsnip_t *pkt,
+                             ng_ipv6_hdr_t *ipv6, uint8_t icmpv6_type,
+                             ng_ndp_opt_t *tl2a_opt, ng_ipv6_addr_t *tgt,
+                             uint8_t adv_flags);
+
+/* send address resolution messages */
+static void _send_nbr_sol(kernel_pid_t iface, ng_ipv6_addr_t *tgt,
+                          ng_ipv6_addr_t *dst);
+static void _send_nbr_adv(kernel_pid_t iface, ng_ipv6_addr_t *tgt,
+                          ng_ipv6_addr_t *dst, bool supply_tl2a);
+
+static void _set_state(ng_ipv6_nc_t *nc_entry, uint8_t state);
+
+/* special netapi helper */
+static inline void _send_delayed(vtimer_t *t, timex_t interval, ng_pktsnip_t *pkt)
+{
+    vtimer_set_msg(t, interval, ng_ipv6_getpid(), NG_NETAPI_MSG_TYPE_SND, pkt);
+}
+
+/* packet queue node allocation */
+static ng_pktqueue_node_t *_alloc_pkt_node(ng_pktsnip_t *pkt);
+static inline void _free_pkt_node(ng_pktqueue_node_t *node);
+
+void ng_ndp_nbr_sol_handle(kernel_pid_t iface, ng_pktsnip_t *pkt,
+                           ng_ipv6_hdr_t *ipv6, ng_ndp_nbr_sol_t *nbr_sol,
+                           size_t icmpv6_size)
+{
+    uint16_t opt_offset = 0;
+    uint8_t *buf = ((uint8_t *)nbr_sol) + sizeof(ng_ndp_nbr_sol_t);
+    ng_ipv6_addr_t *tgt;
+
+    /* check validity */
+    if ((ipv6->hl != 255) || (nbr_sol->code != 0) ||
+        (icmpv6_size < sizeof(ng_ndp_nbr_sol_t)) ||
+        ng_ipv6_addr_is_multicast(&nbr_sol->tgt) ||
+        (ng_ipv6_addr_is_unspecified(&ipv6->src) &&
+         ng_ipv6_addr_is_solicited_node(&ipv6->dst))) {
+        DEBUG("ndp: neighbor solicitation was invalid.\n");
+        /* ipv6 releases */
+        return;
+    }
+
+    if ((tgt = ng_ipv6_netif_find_addr(iface, &nbr_sol->tgt)) == NULL) {
+        DEBUG("ndp: Target address is not to interface %" PRIkernel_pid "\n",
+              iface);
+        /* ipv6 releases */
+        return;
+    }
+
+    icmpv6_size -= sizeof(ng_ndp_nbr_sol_t);
+
+    while (icmpv6_size > 0) {
```

Comment:
I use `int` now… should not lead to problems, since `icmpv6_size` is expected to be `<< 1280`